### ISSUE TITLE
DSSM returning NaN for loss when used with tensorflow-gpu backend.

### ISSUE THREAD

**Issue by MichaelStarkey** [NTMC-Community/MatchZoo#493]
I have been running DSSM on quite a large dataset and was looking at tensorflow-gpu to speed up the training. However the returning loss and mae are always NaN for both the train and evaluation phase. I have tried a very basic tensorflow model from their tutorials and it works fine.

Im not really sure where to start debugging with this, any help would be greatly appreciated.

The model works fine with the cpu version of tensorflow.
Example:
```
model.fit(x,y, epochs=2)
```
```
Epoch 1/2
10000/10000 [==============================] - 1s 139us/step - loss: nan - mean_absolute_error: nan
Epoch 2/2
10000/10000 [==============================] - 1s 138us/step - loss: nan - mean_absolute_error: nan
```

**Comment by faneshion** [comment]
Until now, we have not tested the 2.0 version under the GPU environment. We will test it as soon as possible. If you have any progress, please inform us. Thanks.

**Comment by uduse** [comment]
Are you having the same problem when doing the same thing on a CPU?

**Comment by faneshion** [comment]
Hi, @MichaelStarkey. I have just tested the 2.0 version with GPU environment, and it works fine. Can you give more information about your error?

**Comment by bwanglzu** [comment]
I guess this is an optimization issue, use another loss fucntion.

**Comment by MichaelStarkey** [comment]
It does seem to be an issue with the loss function, as using the hinge losses from keras works ok. MSE and MAE are pretty standard losses however, what would be causing them to not work on the GPU? This is still using DSSM as it comes in MatchZoo.

**Comment by bwanglzu** [comment]
@MichaelStarkey can you try to change the kernel to `tanh` ? [reference](https://github.com/keras-team/keras/issues/1244)

**Comment by MichaelStarkey** [comment]
@bwanglzu 
I think the DSSM model in MatchZoo is using that already.
```
print(str(model.params))
```
```
name                          DSSMModel
model_class                   <class 'matchzoo.models.dssm_model.DSSMModel'>
input_shapes                  [(9645,), (9645,)]
task                          <matchzoo.tasks.ranking.Ranking object at 0x7f577a8b4668>
optimizer                     adam
w_initializer                 glorot_normal
b_initializer                 zeros
dim_fan_out                   128
dim_hidden                    300
activation_hidden             tanh
num_hidden_layers             2
```

Is the activation function what you are referencing?

**Comment by bwanglzu** [comment]
@MichaelStarkey are u using the 2.0-dev branch?

**Comment by MichaelStarkey** [comment]
@bwanglzu I was until I encountered this problem, then switched to 2.0. I had the same problem on both.

**Comment by bwanglzu** [comment]
@MichaelStarkey please use 2.0-dev, can you upload the notebook to somewhere? Let me take a look.

**Comment by bwanglzu** [comment]
if you're reading [this tutorial](https://github.com/NTMC-Community/MatchZoo/blob/2.0-dev/notebooks/tutorial.ipynb), please check one thing: your data does not cotains nan values.

after preprocessing:

```python
from matchzoo import preprocessor
dssm_preprocessor = preprocessor.DSSMPreprocessor()
datapack_train = dssm_preprocessor.fit_transform(train)
```
Please check:
```python
# access three pandas dataframes inside the datapack object.
left = datapack_train.left
right = datapack_train.right
relation = datapack_train.relation
# check whether exist nan values
left.isnull().any()
right.isnull().any()
relation.isnull().any()
```
This might be a reason for the issue.  Possible reasons:

> try different optimizers, f.e. sgd, nadam, adam...
scale you data differently, f.e. try these ranges [0,1] or [-1,1],
Also, in my case the learning rate parameter was the critical one.
and the most important thing:
always check for NaNs or inf in your dataset.
You can do it like this:

[Reference](https://github.com/keras-team/keras/issues/2134)

**Comment by MichaelStarkey** [comment]
Thanks for your help so far, im pretty sure having NaNs in the dataset is not the issue as the model can train completely on the CPU. I uploaded an mvp to my github [here](https://github.com/MichaelStarkey/MatchZooExperiment/).

**Comment by bwanglzu** [comment]
@faneshion @uduse @pl8787  can you take a look at this issue? I need to pack my stuff tonight and fly to China tomorrow.

**Comment by uduse** [comment]
Let's rule out other possible factors first:
- make sure it's not tensorflow version problem by matching your tensorflow version number vs tensorflow-gpu version number
- try converting Y to float numbers, I read your code and it seems they are integers

**Comment by MichaelStarkey** [comment]
The tensorflow and tensorflow-gpu are both on the same version (1.12.0). I also changed the labels to be floats but that unfortunately had no effect.

**Comment by uduse** [comment]
Hm... I don't have a GPU so all suggestions I could give is to rule out factors one by one. If you can reproduce the problem on a sharable GPU, like using a Google Colab, then I might investigate deeper.

**Comment by MichaelStarkey** [comment]
After some experimentation it looks the like problem was entirely caused by the data generator I had made. Sorry for bother and thanks.
